### PR TITLE
Do not store external wallet URLs in Rewards prefs

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/30828

- Creates a new wallet provider abstract method, `AssignWalletLinks`.
- Moves all external wallet links out of prefs.
- Introduces `RewardEngineImpl::GetExternalWalletProvider` in order to centralize the "string-to-wallet-provider" mapping logic.
- Moves `RewardsLogStream` into its own file, and implements `RewardsEngineImpl::Log` and `LogError` in order to provide this logging API to more consumers.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/core/wallet/wallet_balance.cc [hunk 1]
```diff
@@ -64,16 +45,21 @@ void WalletBalance::OnGetUnblindedTokens(
...
-                     base::Unretained(this), wallet_type, std::move(balance),
-                     std::move(callback)));
+  auto* provider = engine_->GetExternalWalletProvider(wallet_type);
+  if (!provider) {
+    BLOG(0, "Invalid external wallet type: " << wallet_type);
+    return std::move(callback).Run(std::move(balance));
+  }
+
+  provider->FetchBalance(base::BindOnce(
+      &WalletBalance::OnFetchExternalWalletBalance, base::Unretained(this),
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/wallet/wallet_balance.cc:61]: This is movement of existing code; all new code will use weak pointers for callbacks.

File: components/brave_rewards/core/wallet_provider/wallet_provider.h [hunk 2]
```diff
@@ -39,6 +39,8 @@ class WalletProvider {
 
   virtual base::TimeDelta GetDelay() const;
 
+  virtual void AssignWalletLinks(mojom::ExternalWallet& external_wallet) = 0;
```
Comment (szilardszaloki) [inline components/brave_rewards/core/wallet_provider/wallet_provider.h:42]: Should we have this as a non-`virtual` function, and have `GetAccountUrl()`, and `GetActivityUrl()` as `virtual` member functions of `WalletProvider` instead?

**Review by szilardszaloki — APPROVED** [review]
LGTM — should we implement a state migration that removes the now-unused wallet provider URLs from `Preferences`?

File: components/brave_rewards/core/wallet_provider/wallet_provider.h (hunk 2 above)
Comment (zenparsing) [reply components/brave_rewards/core/wallet_provider/wallet_provider.h:42]: I can see arguments both ways here. I think for now though we should opt for the simpler API (a single `AssignWalletLinks` method).

**Comment by zenparsing** [comment]
Regarding state migrations: the old stored URLs will naturally disappear the next time the wallet info is saved, which should occur on regular "reauth" intervals.

**Comment by szilardszaloki** [comment]
Ah, of course — makes sense! 👍🏻

**Pull request merged**